src/libraries/System.Private.CoreLib/src/System/SpanHelpers.cs

```diff
@@ -12,323 +13,88 @@ namespace System
-                    Unsafe.As<byte, int>(ref Unsafe.Add(ref b, 8)) = 0;
-                    Unsafe.As<byte, int>(ref Unsafe.Add(ref b, 12)) = 0;
-#endif
-                    Unsafe.As<byte, int>(ref Unsafe.Add(ref b, 16)) = 0;
-                    return;
-                case 21:
-#if TARGET_64BIT
-                    Unsafe.As<byte, long>(ref b) = 0;
-                    Unsafe.As<byte, long>(ref Unsafe.Add<byte>(ref b, 8)) = 0;
-#else
-                    Unsafe.As<byte, int>(ref b) = 0;
-                    Unsafe.As<byte, int>(ref Unsafe.Add(ref b, 4)) = 0;
-                    Unsafe.As<byte, int>(ref Unsafe.Add(ref b, 8)) = 0;
-                    Unsafe.As<byte, int>(ref Unsafe.Add(ref b, 12)) = 0;
-#endif
-                    Unsafe.As<byte, int>(ref Unsafe.Add(ref b, 16)) = 0;
-                    Unsafe.Add(ref b, 20) = 0;
-                    return;
-                case 22:
-#if TARGET_64BIT
-                    Unsafe.As<byte, long>(ref b) = 0;
-                    Unsafe.As<byte, long>(ref Unsafe.Add<byte>(ref b, 8)) = 0;
-#else
-                    Unsafe.As<byte, int>(ref b) = 0;
-                    Unsafe.As<byte, int>(ref Unsafe.Add(ref b, 4)) = 0;
-                    Unsafe.As<byte, int>(ref Unsafe.Add(ref b, 8)) = 0;
-                    Unsafe.As<byte, int>(ref Unsafe.Add(ref b, 12)) = 0;
-#endif
-                    Unsafe.As<byte, int>(ref Unsafe.Add(ref b, 16)) = 0;
-                    Unsafe.As<byte, short>(ref Unsafe.Add(ref b, 20)) = 0;
-                    return;
-            }
-
-            // P/Invoke into the native version for large lengths
-            if (byteLength >= 512) goto PInvoke;
-
-            nuint i = 0; // byte offset at which we're copying
 
-            if (((nuint)Unsafe.AsPointer(ref b) & 3) != 0)
+            if (Vector.IsHardwareAccelerated && byteLength >= (uint)(Vector<byte>.Count))
             {
-                if (((nuint)Unsafe.AsPointer(ref b) & 1) != 0)
+                // We have enough data for at least one vectorized write.
+                nuint stopLoopAtOffset = byteLength & (nuint)(nint)(2 * (int)-Vector<byte>.Count); // intentional sign extension carries the negative bit
+                nuint offset = 0;
+
+                // Loop, writing 2 vectors at a time.
+                // Compare 'numElements' rather than 'stopLoopAtOffset' because we don't want a dependency
+                // on the very recently calculated 'stopLoopAtOffset' value.
+                if (byteLength >= (uint)(2 * Vector<byte>.Count))
                 {
-                    b = 0;
-                    i += 1;
-                    if (((nuint)Unsafe.AsPointer(ref b) & 2) != 0)
-                        goto IntAligned;
+                    do
+                    {
+                        Vector<byte>.Zero.StoreUnsafe(ref b, offset);
+                        Vector<byte>.Zero.StoreUnsafe(ref b, offset + (nuint)Vector<byte>.Count);
```

Comment:
```suggestion
                        Unsafe.AddByteOffset(ref Unsafe.As<byte, (Vector<byte>, Vector<byte>)>(ref b), offset) = default;
```

Reply:
I am not entirely sure tuples will be handled well by mono runtime, also, IMO I find it less clear to read 😐

Reply:
We can avoid the redundant vxorps ymm0, ymm0, ymm0.

Reply:
I think we'd better fix this in JIT 🙂